### PR TITLE
fix(swap): display very small numbers max values, fix scientific notation, display long number & handle decimal mismatch

### LINKED ISSUE #21330: Handling decimal mismatch when switching tokens with different decimal support
### Steps to Reproduce:
1. Select a token with 18 decimal support (like DAI) in the Send field.
2. Change the token to one with 6 decimal support (like USDC).

### Actual Result:
The behavior is inconsistent or undefined when switching tokens with different decimal places.

https://github.com/user-attachments/assets/c74d27fb-2840-4d44-b6ef-1292bc0acab6

### Expected Result:
Potential solutions:
- Clear the field: Some apps, like TrustWallet and Binance, clear the value to 0 when switching tokens with different decimal support.

https://github.com/user-attachments/assets/b02496b1-0b25-4f74-894a-aef006ad2340

OR

- Other apps, like Uniswap, do not treat this as a critical issue. They allow the 18 decimals to remain for tokens with only 6 decimal support and do not recalculate the value. 

https://github.com/user-attachments/assets/b2f426ee-6148-4fe7-b25e-970e05445195

In such case the validation message "Token supports fewer decimals. Please adjust the value." should be shown

discussed with design team https://discord.com/channels/1210237582470807632/1270072144067563520/1288802954626990090

### OS: 
IOS, Android
### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### LINKED ISSUE #21331: Default token in the receiving field of swap is not SNT
### Steps:
1. Open the swap page.
2. Navigate to the swap page selecting any token
3. Check the default token in the "Receive" field.

### Actual result:
A random token is shown by default in the receiving field of the swap

https://github.com/user-attachments/assets/f40294b4-e6fe-453b-8b58-cf943ca323f8

### Expected result:
- SNT should be shown as the default token in the receiving field.
- If SNT is selected as the token to send, the receive field should be empty.

image

https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=43-83983&node-type=instance&m=dev
discussed with https://discord.com/channels/1210237582470807632/1270072144067563520/1288121240757862553

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### LINKED ISSUE #21347: Long numbers in swap receive field not properly displayed
### Steps:
1. Go to swap
2. Select some 18 decimals token in the receiving value

### Actual result:
- If the number is long, hidden digits are not blurred.
- The number is not shown from the beginning in the receive field.
image

### Expected result:

- Show the swap as per design. The  numbers should be blurred in case if some numbers are hidden
- The the number is shown from the begining in the receive field
image

### LINKED ISSUE #21348: Remaining decimals token shown in Scientific Notation on swap page if max label is tapped
### Steps to Reproduce:
1. Restore a user with a token that has a very small remaining decimal value (e.g., 0.000000000262783357 DAI).
2. Go to the swap page and select this token.
3. Tap "Max" value.

### Actual Result:
The token value is shown in scientific notation.

image

### Expected Result:
The value should be displayed in real number format, not in scientific notation.

### Devices:
Pixel 7a, Android 13
iPhone 11 Pro Max, iOS 17

### ENV:
Nightly 30.09, 2024

### LINKED ISSUE #21405: The predefined asset in the 'Receive asset' field is set to SNT, when SNT is selected as the 'Asset to Pay,'
### Steps:
1. Go to the list of assets
2. Select SNT asset
3. Navigate to swap

### Actual result:
The predefined asset in the 'Receive asset' field is set to SNT, when SNT is selected as the 'Asset to Pay,'

### Expected result:
If SNT is selected as the asset to pay, the predefined asset in the receive field should be ETH. 

discussed in https://discord.com/channels/1210237582470807632/1270072144067563520/1293272307904806912
same issue is created on desktop side https://github.com/status-im/status-desktop/issues/16494

### PR DESCRIPTION
**Pull request by briansztamfater**
fixes #21348
fixes #21347
fixes #21331
fixes #21330
fixes #21405

### Summary

This PR fixes a couple of bugs in the Swap flow:
1. Remaining decimals token shown in Scientific Notation on swap page if max label is tapped
2. Long numbers in swap receive field not properly displayed
3. Default token in the receiving field of swap is not SNT
4. Handling decimal mismatch when switching tokens with different decimal support

#### Platforms

- Android
- iOS

##### Functional

- wallet / transactions

#### Issue 1
- Open Status
- Restore a user with a token that has a very small remaining decimal value (e.g., 0.000000000262783357 DAI).
- Go to swaps
- Select that token in the select asset to pay screen
- Tap "Max" value
- Verify the value is not in scientific notation

#### Issue 2
- Open Status
- Log in
- Go to swaps
- Select some 18 decimals token in the receiving value
- Enter some values
- Verify the values are displayed properly with gradients in the beginning of the input (pay token) and in the end pf the input (receive token)

[NOTE] There's an Android bug in our current version of React Native (0.73) that prevents us to position the text input at the start of it (https://github.com/facebook/react-native/issues/44171), thus will have to follow-up on this once we update to 0.74+ where this seems to be fixed (https://github.com/reactwg/react-native-releases/issues/283)

#### Issue 3 
- Open Status
- Login
- Go to swaps
- Select an asset
- Verify the default asset to receive is SNT

#### Issue 4 
- Open Status
- Login
- Go to swaps
- Select a token with 18 decimal support (like DAI) in the asset to pay field
- Change the token to one with 6 decimal support (like USDC)
- Verify the input is cleared when decimals of the current value and thew new selected token don't match

status: ready

### CODE REVIEW

**Review by mohsen-ghafouri — APPROVED** [review]
LGTM

File: src/quo/components/wallet/swap_input/style.cljs [hunk 1]
```diff
@@ -75,3 +75,19 @@
...
+   :bottom   0
+   :z-index  1})
+
+(def gradient-end
+  {:width    64
+   :position :absolute
+   :top      0
+   :right    0
+   :bottom   0
+   :z-index  1})
```
Comment (seanstrom) [inline src/quo/components/wallet/swap_input/style.cljs]: Maybe we can add a `def` named `gradient-common`  for all the shared keys?

File: src/quo/components/wallet/swap_input/view.cljs [hunk 2]
```diff
@@ -42,26 +44,65 @@
       [:container-style {:optional true} [:maybe :map]]]]]
    :any])
 
+(def icon-size 32)
+(def container-padding 24)
+(def max-cursor-position 5)
```
Comment (seanstrom) [inline src/quo/components/wallet/swap_input/view.cljs]: How are we calculating the `container-padding` value? Is it possible to reference this from a style file?

Comment (seanstrom) [reply src/quo/components/wallet/swap_input/view.cljs]: for `max-cursor-position`, is it possible that value will need to smaller on devices with smaller screens?

**Review by seanstrom — COMMENTED** [review]
Thanks for the PR 🙌 
Overall, I think the code looks good, but I'm just not sure I understand some of the intended behaviour, so I left a few comments.

If possible, could you also provide some screen captures showing the new behaviour please? 🙏

**Comment by VolodLytvynenko** [comment]
hi @briansztamfater thank you for PR.
[issue 21331](https://github.com/status-im/status-mobile/issues/21348) is fixed
[issue 21348](https://github.com/status-im/status-mobile/issues/21347) still reproducible for Android devices. First numbers are not visible in the "receive" field
Pixel XL API 33 (Android studion)
image

Pixel 7a, Android 13
image

**Comment by briansztamfater** [comment]
Hey @VolodLytvynenko, that's a known issue, check note from issue 2 in PR description. We can create a issue to follow up after we upgrade the React Native version

**Comment by VolodLytvynenko** [comment]
> Hey @VolodLytvynenko, that's a known issue, check note from issue 2 in PR description. We can create a issue to follow up after we upgrade the React Native version

@briansztamfater Oh yes, sorry for this. I have missed this

Unfortunately [issue 21331](https://github.com/status-im/status-mobile/issues/21331) is not fixed fully. There is some additional way to make SNT not default in the receive field. Take a look please

### Steps to Reproduce:
1. User is on the main page.
2. Long tap the asset and tap "Swap."
### Actual Result:
SNT is not shown as the default token in the "receive" field.

https://github.com/user-attachments/assets/10d80704-e266-438d-9a45-47a5456b366f

### Expected result
SNT is shown as the default token in "receive" field

**Comment by briansztamfater** [comment]
> > Hey @VolodLytvynenko, that's a known issue, check note from issue 2 in PR description. We can create a issue to follow up after we upgrade the React Native version
> 
> @briansztamfater Oh yes, sorry for this. I have missed this
> 
> Unfortunately [issue 21331](https://github.com/status-im/status-mobile/issues/21331) is not fixed fully. There is some additional way to make SNT not default in the receive field. Take a look please
> 
> ### Steps to Reproduce:
> 1. User is on the main page.
> 2. Long tap the asset and tap "Swap."
> 
> ### Actual Result:
> SNT is not shown as the default token in the "receive" field.
> 
>  snt.mp4 
> ### Expected result
> SNT is shown as the default token in "receive" field

@VolodLytvynenko thanks for testing, and sorry about that, issue should be fixed now :)

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko thanks for testing, and sorry about that, issue should be fixed now :)

Hey, thank you for the fix! SNT is usually shown as the default, but sometimes it doesn’t appear, especially right after the initial login

### Steps to Reproduce:
1. User is on the main page.
2. Long tap the asset and tap "Swap."
3. Check default asset
4. Relogin and try again steps 1-3 if issue is not reproduced

### Actual result:
SNT is not shown as the default token in the "Receive" field after relogin.

https://github.com/user-attachments/assets/08c75141-8426-42d9-a83d-61662ce26819

### Expected Result:
SNT should be displayed as the default token in the "Receive" field after relogin.

### ENV:
IOS, Android
[logs.zip](https://github.com/user-attachments/files/17306056/logs.zip)

**Comment by VolodLytvynenko** [comment]
@briansztamfater additionally, we [discussed](https://discord.com/channels/1210237582470807632/1270072144067563520/1293226718882168962) the issue related to predefined SNT yesterday. Could you take a look and see if this [issue](https://github.com/status-im/status-mobile/issues/21405) can be fixed within the scope of this PR? Otherwise, it can be addressed separately

**Comment by churik** [comment]
@briansztamfater @VolodLytvynenko if issue is really hard to fix, we can postpone it but the preference is to have it fixed in Pr scope

**Comment by briansztamfater** [comment]
@VolodLytvynenko @churik no problem at all, both issues should be fixed now!

**Comment by VolodLytvynenko** [comment]
hi @briansztamfater thank you for your work. PR is ready to be merged

File: src/quo/components/wallet/swap_input/view.cljs (hunk 2 above)
Comment (briansztamfater) [reply src/quo/components/wallet/swap_input/view.cljs]: > for `max-cursor-position`, is it possible that value will need to smaller on devices with smaller screens?

It is highly unlikely as 5 is a very low value, also this handle a very uncommon edge case when the user moves the cursor of the input and the value is a very long number, so I wouldn't mind for now.

> How are we calculating the `container-padding` value? Is it possible to reference this from a style file?

It is the sum of left and right padding. Updated the code to extract the values from the styles!

File: src/quo/components/wallet/swap_input/style.cljs (hunk 1 above)
Comment (briansztamfater) [reply src/quo/components/wallet/swap_input/style.cljs]: Thanks for the suggestion, updated :)

**Pull request merged**